[JIT] Fix function schema subtype checking

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47706 [JIT] Fix function schema subtype checking**

**Summary**
This commit fixes `FunctionSchema::isSubtypeOf` so that the subtyping rule it
implements for `FunctionSchema` instances is contravariant in argument
types and covariant in return type. At present, the rule is covariant in
argument types and contravariant in return type, which is not correct.

A brief but not rigourous explanation follows. Suppose there are two
`FunctionSchema`s, `M = (x: T) -> R` and `N = (x: U) -> S`. For `M <= N`
to be true (i.e. that `M` is a subtype of `N`), it must be true that `U
<= T` and `R <= S`. This generalizes to functions with multiple
arguments.

**Test Plan**
This commit extends `TestModuleInterface.test_module_interface_subtype`
with two new tests cases that test the contravariance of argument types
and covariance of return types in determining whether a `Module`
implements an interface type.

**Fixes**
This commit closes #47631.

Differential Revision: [D24934099](https://our.internmc.facebook.com/intern/diff/D24934099)

Commit:
[JIT] Fix function schema subtype checking

**Summary**
This commit fixes `FunctionSchema::isSubtypeOf` so that the subtyping rule it
implements for `FunctionSchema` instances is contravariant in argument
types and covariant in return type. At present, the rule is covariant in
argument types and contravariant in return type, which is not correct.

A brief but not rigourous explanation follows. Suppose there are two
`FunctionSchema`s, `M = (x: T) -> R` and `N = (x: U) -> S`. For `M <= N`
to be true (i.e. that `M` is a subtype of `N`), it must be true that `U
<= T` and `R <= S`. This generalizes to functions with multiple
arguments.

**Test Plan**
This commit extends `TestModuleInterface.test_module_interface_subtype`
with two new tests cases that test the contravariance of argument types
and covariance of return types in determining whether a `Module`
implements an interface type.

**Fixes**
This commit closes #47631.

[ghstack-poisoned]

gmagogsfm commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -288,12 +288,12 @@ inline bool FunctionSchema::isSubtypeOf(
     std::ostream* why_not) const {
   size_t start = as_method ? 1 : 0;
-  // functions are covariant in arguments but contravariant in returns
+  // functions are contravariant in arguments but covariant in returns
```
Now I am curious how things didn't completely collapse with previous incorrect implementation ...

SplitInfinity replied:
Me too!

gmagogsfm commented on test/jit/test_module_interface.py:
```diff
@@ -202,6 +202,40 @@ def forward(self, x):
+
+        as_tensor_to_any(torch.jit.script(TensorToAnyImpl()))
+        as_any_to_any(torch.jit.script(AnyToAnyImpl()))
+
```
Optional:

It seems we don't have a test case where parameters are contravariant *and* results are covariant at the same time. I think it can be done via having a TensorToAnyImpl with f: Any->Tensor.

gmagogsfm reviewed (approved):
Nice!

Commit:
Update on "[JIT] Fix function schema subtype checking"


**Summary**
This commit fixes `FunctionSchema::isSubtypeOf` so that the subtyping rule it
implements for `FunctionSchema` instances is contravariant in argument
types and covariant in return type. At present, the rule is covariant in
argument types and contravariant in return type, which is not correct.

A brief but not rigourous explanation follows. Suppose there are two
`FunctionSchema`s, `M = (x: T) -> R` and `N = (x: U) -> S`. For `M <= N`
to be true (i.e. that `M` is a subtype of `N`), it must be true that `U
<= T` and `R <= S`. This generalizes to functions with multiple
arguments.

**Test Plan**
This commit extends `TestModuleInterface.test_module_interface_subtype`
with two new tests cases that test the contravariance of argument types
and covariance of return types in determining whether a `Module`
implements an interface type.

**Fixes**
This commit closes #47631.

[ghstack-poisoned]

jeffdaily commented:
ROCm CI failure due to CI script update.  Will retest.

jeffdaily commented:
@pytorchbot retest this please
